what4/src/What4/Utils/Serialize.hs

```diff
@@ -0,0 +1,132 @@
+{-# LANGUAGE NondecreasingIndentation #-}
+{-# LANGUAGE ConstraintKinds #-}
+{-# LANGUAGE DataKinds #-}
+{-# LANGUAGE FlexibleContexts, FlexibleInstances #-}
+{-# LANGUAGE GADTs #-}
+{-# LANGUAGE KindSignatures #-}
+{-# LANGUAGE LambdaCase #-}
+{-# LANGUAGE PolyKinds #-}
+{-# LANGUAGE RankNTypes #-}
+{-# LANGUAGE ScopedTypeVariables #-}
+{-# LANGUAGE TypeApplications #-}
+{-# LANGUAGE TypeFamilies #-}
+{-# LANGUAGE AllowAmbiguousTypes #-}
+{-# LANGUAGE TypeOperators #-}
+module What4.Utils.Serialize
+    (
+      withRounding
+    , fromJust'
+    , makeSymbol
+    , SomeSome(..)
+    , asyncLinked
+    , withAsyncLinked
+    ) where
+
+import qualified Control.Exception as E
+import           Data.Kind
+import           Text.Printf ( printf )
+import           Data.Maybe ( fromMaybe )
+import qualified Data.BitVector.Sized as BV
+import           What4.BaseTypes
+import           What4.Serialize.Log
+import qualified What4.Interface as S
+import           What4.Symbol ( SolverSymbol, userSymbol )
+
+
+import qualified UnliftIO as U
+
+----------------------------------------------------------------
+-- * Async
+
+-- | Fork an async action that is linked to the parent thread, but can
+-- be safely 'U.cancel'd without also killing the parent thread.
+--
+-- Note that if your async doesn't return unit, then you probably want
+-- to 'U.wait' for it instead, which eliminates the need for linking
+-- it. Also, if you plan to cancel the async near where you fork it,
+-- then 'withAsyncLinked' is a better choice than using this function
+-- and subsequently canceling, since it ensures cancellation.
+--
+-- See https://github.com/simonmar/async/issues/25 for a perhaps more
+-- robust, but also harder to use version of this. The linked version
+-- is harder to use because it requires a special version of @cancel@.
+asyncLinked :: (U.MonadUnliftIO m) => m () -> m (U.Async ())
+asyncLinked action = do
+  -- We use 'U.mask' to avoid a race condition between starting the
+  -- async and running @action@. Without 'U.mask' here, an async
+  -- exception (e.g. via 'U.cancel') could arrive after
+  -- @handleUnliftIO@ starts to run but before @action@ starts.
+  U.mask $ \restore -> do
+  a <- U.async $ handleUnliftIO (\E.ThreadKilled -> return ()) (restore action)
```

Comment:
This doesn't seem like the most up-to-date version of the code from `what4-serialize`, and the corresponding upstream code [here](https://github.com/GaloisInc/what4-serialize/blob/d8b3fa73f7001614c593792aa082e13f4191e110/src/What4/Utils/Util.hs#L60) looks different. More pressingly, this version of the code emits warnings on GHC 9.2, as can be seen in the CI.

Similar comments apply to the rest of the code changed in [this commit](https://github.com/GaloisInc/what4-serialize/commit/3b0a1497bee928d6cb288248e6b91633fced5fff).

Reply:
Whoops, I think I fixed that now. Good catch. My local copy was embarrassingly out of date